docs/proposals/compute-resource-metrics-api.md

```diff
@@ -0,0 +1,195 @@
+# Kubernetes compute resource metrics API
+
+## Goals
+
+Provide resource usage metrics on pods and nodes on the API server to be used
+by the scheduler to improve job placement, utilization, etc. and by end users
+to understand the resource utilization of their jobs.
+
+## Current state
+
+Right now, the Kubelet exports container metrics via an API endpoint. This
+information is not gathered nor served by the Kubernetes API server.
+
+## Use cases
+
+The first user will be kubectl. The resource usage data can be shown to the
+user via a periodically refreshing interface similar to `top` on Unix-like
+systems. This info could let users assign resource limits more efficiently.
+
+```
+$ kubectl top kubernetes-minion-abcd
+Pod: monitoring-heapster-abcde
+CPU:
+    mean: 0.12 cores
+    95th: 1.10 cores
+    max:  1.91 cores
+Memory:
+    mean: 300MB (10%)
+    95th: 330MB (11%)
+    max:  395MB (13%)
+```
+
+A second user will be the scheduler. To assign pods to nodes efficiently, the
+scheduler needs to know the current free resources on each node.
+
+## Proposed endpoints
+
+	/api/v1/namespaces/myns/podMetrics/mypod/[derived|raw]
+	/api/v1/nodeMetrics/myNode/[derived|raw] 
+
+The derived metrics include the mean, max and a few percentiles of the list of
+values.
+
+The raw metrics are the values themselves, thus the data will be much heavier.
+
+In the future, if any uses cases are found that would benefit from RC,
+namespace or service aggregation, metrics at those levels could also be
+exposed taking advantage of the fact that Heapster already does aggregation
+and metrics for them.
+
+## Data gathering
+
+We can either use a Push or a Poll based system:
+
+* Push - each node would periodically POST their derived and raw metrics to
+  the API server. Kubectl and the scheduler would watch the API server to get
+  the latest metrics.
+
+* Poll - heapster would periodically poll a kubelet API endpoint on each node
+  to fetch its node and pods metrics. Then Heapster will have to implement the
+  APIs mentioned above.
+
+Our initial plan is to follow the push based system.
+
+The API server will not hold any of this data directly. For our initial
+purposes, it will hold the most recent metrics obtained from each node in
+etcd. Then, when polled for metrics, the API server would only serve said most
```

Comment:
Or optionally, just call through to the underlying location of the metrics
(node today, a metrics store like heapster later)?  It seems like you only
want metrics as a snapshot, and historical / more advanced metrics would be
elsewhere.  That sounds a lot like how we treat pod logs today (we call to
the api server which calls to the kubelet).

On Tue, Jul 28, 2015 at 10:49 PM, Eric Tune <notifications@github.com>
wrote:

> In docs/proposals/compute-resource-metrics-api.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11951#discussion_r35723817>
> :
>
> > +
> > +We can either use a Push or a Poll based system:
> > +
> > +* Push - each node would periodically POST their derived and raw metrics to
> > +  the API server. Kubectl and the scheduler would watch the API server to get
> > +  the latest metrics.
> > +
> > +* Poll - heapster would periodically poll a kubelet API endpoint on each node
> > +  to fetch its node and pods metrics. Then Heapster will have to implement the
> > +  APIs mentioned above.
> > +
> > +Our initial plan is to follow the push based system.
> > +
> > +The API server will not hold any of this data directly. For our initial
> > +purposes, it will hold the most recent metrics obtained from each node in
> > +etcd. Then, when polled for metrics, the API server would only serve said most
>
> Can we skip putting it in etcd? Otherwise it will spam the etcd window and
> logs, I think? Since it will be recreated within a few seconds, no need to
> persist it. Just need some way to do unreliable broadcast from one
> apiserver to the other replicas?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11951/files#r35723817>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift